protowhat/sct_syntax.py

```diff
@@ -1,134 +1,329 @@
+                + ", ".join(
+                    chain_iters(
+                        (str(arg) for arg in self.args),
+                        (
+                            "{}={}".format(kwarg, value)
+                            for kwarg, value in self.kwargs.items()
+                        ),
+                    )
+                )
+                + ")"
+            )
+
+
 class Chain:
-    def __init__(self, state, attr_scts=None):
-        self._state = state
-        self._crnt_sct = None
-        self._waiting_on_call = False
-        self._attr_scts = {} if attr_scts is None else attr_scts
+    empty_call_str = ""
+
+    def __init__(
+        self,
+        chained_call: Optional[Callable] = None,
+        previous: Optional["Chain"] = None,
+        chainable_functions: Dict[str, Callable] = None,
+    ):
+        if not chained_call and previous:
+            raise ValueError("After the start of a chain a call is required")
+
+        self.call = chained_call
+        self.previous = previous
+        self.next = []
+
+        if self.previous:
+            previous.next.append(self)
+
+        if chainable_functions:
+            self.chainable_functions = chainable_functions
+        elif previous:
+            self.chainable_functions = previous.chainable_functions
+        else:
+            self.chainable_functions = {}
 
-    def _double_attr_error(self):
-        raise AttributeError(
-            "Did you forget to call a statement? "
-            "e.g. Ex().check_list_comp.check_body()"
-        )
+    @property
+    def _history(self) -> list:
+        return getattr(self.previous, "_history", []) + [self]
 
     def __getattr__(self, attr):
-        # Enable fast attribute access
-        if attr == "_attr_scts":
-            raise AttributeError("Prevent getattr recursion on copy")
-        attr_scts = self._attr_scts
-        if attr not in attr_scts:
-            raise AttributeError("No SCT named %s" % attr)
-        elif self._waiting_on_call:
-            self._double_attr_error()
+        chainable_functions = self.chainable_functions
+        if attr not in chainable_functions:
+            if attr not in ("_history",):  # todo
+                raise AttributeError("No function named %s" % attr)
+            return self.__getattribute__(attr)
         else:
-            # make a copy to return,
-            # in case someone does: chain = chain_of_checks; chain.a; chain.b
-            return self._sct_copy(attr_scts[attr])
-
-    def __call__(self, *args, **kwargs):
-        self._state = link_to_state(self._crnt_sct)(self._state, *args, **kwargs)
-        self._waiting_on_call = False
-        return self
-
-    def __rshift__(self, f):
-        if self._waiting_on_call:
-            self._double_attr_error()
-        elif isinstance(f, Chain) and not isinstance(f, F):
+            # in case someone does: a = chain.a; a(...); a(...)
+            return ChainExtender(self, chainable_functions[attr])
```

Comment:
Why not `return self >> chainable_functions[attr]`?

Reply:
That would return a Chain instead of a ChainExtender